### DISCUSSION TITLE
Building without certain (or all) languages

### DISCUSSION THREAD

**Discussion by CptPotato** [helix-editor/helix#1483] [Q&A] [answered]
As far as I can tell it currently is mandatory to pull in **all** language submodules in order to build helix.
The number of supported languages has increased a lot recently, which makes cloning the repository and building helix take much longer that it used to. (also requiring more disk space to do so)

Is it possible for me to only check out some of the the language submodules and get helix to build?

**Comment by archseer** [comment, marked as answer]
I suggest shallow cloning via `--shallow-submodules`, but you can also go and clone the main repo without submodules, then fetch only the ones you want under `helix-syntax/languages`.

I think in the long run we want to stop pulling in submodules and instead have a separate `cargo xtask` that will take a list of `<url> <sha>`, fetch & compile. That will make distribution packaging a bit more convoluted though.

**Reply by CptPotato** [reply]
What worked for me is the following:

- `git clone https://github.com/helix-editor/helix --shallow-submodules --depth=1`
- `cd helix/helix-syntax/languages/`
- `rm -rv ./*`
- `git submodule update --init --depth=1 <submodule-name>` for each language that you want to build
  - for example: `git submodule update --init --depth=1 tree-sitter-rust`
- go to `helix/helix-syntax/build.rs` and remove these lines if they're not part of your selected languages:
https://github.com/helix-editor/helix/blob/8d273a561387ceb0bcc9c9f8c9fa90a64204118f/helix-syntax/build.rs#L202-L205

Then you can go back to the repo root and install with:
`cargo install --path helix-term`

If the build of a language fails, check out it's required submodules. The lazy way would be checking out all submodules recursively in the steps above. (i.e. add the `--recursive` flag to the `git submodule update` - this might pull in unnecessary repos, though)
